### ISSUE TITLE
TCK: How to test Entity

### ISSUE THREAD

**Issue by KyleAure** [jakartaee/data#119]
This issue builds off of https://github.com/jakartaee/data/issues/112

## Description

Since we've decided that the API for Entity will remain in the Persistence and NoSQL specs where they will be developed separately. 
We still need to figure out a way to write TCK tests that use entities. 

This is tricky since we want to be able to test `jakarta.nosql.Entity` or `jakarta.persistence.Entity` or both during a single run of the TCK.  I've been trying to think of a few ways to do this with the least amount of duplication possible: 

## Main roadblock

The main roadblock here is that Entity is an annotation that needs to exist on an object at compile time. 

## Possible Solutions

Here are some possible solutions that I have thought of, but don't have a complete solution and need help choosing/refining one. 

### Dual Annotated Entities
Add both Entity annotations for example:

```java
@jakarta.nosql.Entity
@jakarta.persistence.Entity
public class Student {
  
  @jakarta.nosql.Id
  @jakarta.persistence.Id
  Long id;
}

@Repository
public interface StudentDirectory extends DataRepository<Student, Long> {
    void save(Student student);
    int countByAgeGreaterThanEqual(Integer age);
}
``` 

For a standalone repository provider or a Jakarta EE profile that only has a single repository provider, this will work.
However, what happens when `studentDirectory.save()` is run on a Jakarta EE profile that has both providers?
How can we run the same test against both providers? 

### Duplicate code
This is a messy but straightforward solution by just duplicating code. For example: 

```java
public class Student {
  @jakarta.persistence.Entity
  public static class Persistence { ...  }

  @jakarta.nosql.Entity
  public static class NoSQL { ... }
}

public interface StudentDirectory {
    @Repository
    public interface Persistence extends DataRepository<Student.Persistence, Long> { ... }

	@Repository
	public interface NoSQL extends DataRepository<Student.NoSQL, Long> { ... }
```

With this solution, we would have to be told by the runner of the TCK which repository(ies) are supported and run based on that input.  This will likely require some more complex Arquillan/Junit logic but should be doable.

### Generated source code
Use a pre-compile tool that will automatically generate source code, compile it, and include it in the TCK.
This will require some sort of parser to search for a mock Entity/Id annotations and use that class as a template, for example: 

```java
@ee.jakarta.tck.data.anno.Entity
public class Student {
  
  @@ee.jakarta.tck.data.anno.Id
  Long id;
}
```

Once parsed two files are created: 
```java
@jakarta.nosql.Entity
public class StudentNoSQL { ... }

@jakarta.persistence.Entity
public class StudentPersistance { ... }
```

The issues with this solution are that it is complicated to implement, it adds complexity that needs to be maintained, and we will have to generate both Entity and Repository classes.  Additionally, at runtime, we will still need to have some method for generating separate applications for both repositories.  Basically, a complex approach to avoid duplicated code.

**Comment by keilw** [comment]
The last one with the generator sounds a little ickie, 
Dual annotated feels a little cleaner.
I recall we used multi-annotations (3 or more) on entities that could be persisted via JPA, JSON (before the Jakarta JSON spec but similar) and through POI even as an Excel sheet.

The approach with two or more embedded entities under the actual object feels rather weird, too.
So IMO 1. sounds best.

**Comment by njr-11** [comment]
My preference would be that we make the TCK look as close to customer code as possible.  I'm fine with duplicating the same entities for Jakarta Persistence and NoSQL.

```
@jakarta.persistence.Entity
public class Student {
  ...
}

@Repository
public interface StudentDirectory extends DataRepository<Student, Long> {
  ...
}

// Do we need a separate package for these, or can we use the same package but compile 2 separate artifacts, one for NoSQL and one for JPA?
@jakarta.nosql.Entity
public class Student {
  ...
}

@Repository
public interface StudentDirectory extends DataRepository<Student, Long> {
  ...
}

```

When running the TCK, you would just specify a mode for NoSQL or JPA to tell it which artifact to include in the application so that there is no runtime dependency on the one you aren't using.  The main issue would be over whether we can have both entities and both repositories have the same class and package name (which would be easier for writing the rest of the tests), or whether they will need to be in separate packages to distinguish.

**Comment by otaviojava** [comment]
Let's increase the challenge a little more:

* It would be fantastic if we could test injecting different DI engine
* We must remember that SQL has features that must be clarified to NoSQL.

Maybe the JPA TCK has inspiration for us on this point.

**Comment by KyleAure** [comment]
I've been doing some testing: 

@keilw - Dual annotation will work for both the Jakarta Profile and Standalone tests.  I've confirmed with the IBM Java team that when Class#getAnnotations()  and Class#getDeclaredAnnotations() are called only annotations that are available to the classloader are returned.  However, we still have the problem of not having a way to ensure a certain repository provider when both are available.

@njr-11 - I looked more into your suggestion of having copies of the entities and a single repository provider. 
The problem I've run into is the compilation.  Compilation of these classes will have to happen prior to the distribution of the TCK and there is no way to compile two classes with the same name in the same package.
Therefore these classes need to come from different packages which will have a cascading effect of duplication: 
- If the entities are in different packages, then we need to duplicate the Repositories.
- If the entities are in different packages, then we also need to duplicate tests since we will need to create instances of the entity for testing.
- If we duplicate the repositories, we need to figure out a mechanism for looking up the correct repository at runtime.

@otaviojava - Not sure what you mean by ` injecting different DI engine`.  For Jakarta EE platforms the dependency injection API will be CDI the implementation used is up to the platform, not the TCK.  For the standalone case, we have not defined a provider method for standalone implementations.  Until that is decided we cannot create tests for the standalone case.

**Comment by KyleAure** [comment]
I created example PRs for the two leading solutions.
dual annotations vs. duplicate entities

**Comment by njr-11** [comment]
Kyle, thanks for putting together the two examples - it is very helpful. Duplicate entities without the ability to compile two copies of the same class certainly ends up being messy and no longer matches how a customer would write code.  Dual annotations is much cleaner.  If you've solved the concern there about unavailablility of classes at runtime, the other issue that will come up is that the TCK becomes vulnerable to challenges that its entities (and thus tests that use them) aren't valid because they have both sets of annotations on them and the spec doesn't require providers to support that.  If we are going to write the TCK to depend on do dual annotations, we need to be willing to put language explicitly into the Jakarta Data specification stating that dual annotations must be allowed to ensure the TCK is valid.  We should discuss this topic more on tomorrow's call.

**Comment by KyleAure** [comment]
This has been resolved with the merge of #120
